tensor2tensor/utils/bleu_hook.py

```diff
@@ -209,7 +210,11 @@ def bleu_wrapper(ref_filename, hyp_filename, case_sensitive=False):
 def _read_stepfiles_list(path_prefix, path_suffix=".index", min_steps=0):
   """Return list of StepFiles sorted by step from files at path_prefix."""
   stepfiles = []
-  for filename in tf.gfile.Glob(path_prefix + "*-[0-9]*" + path_suffix):
+  # tf.gfile.Glob may crash with
+  # tensorflow.python.framework.errors_impl.NotFoundError:
+  # xy/model.ckpt-1130761_temp_9cb4cb0b0f5f4382b5ea947aadfb7a40; No such file or directory
+  # Let's use standard glob.glob instead.
+  for filename in glob.glob(path_prefix + '*-[0-9]*' + path_suffix):
```

Comment:
would prefer tf.gfile.Glob. why is it crashing?

Reply:
When saving a checkpoint, TensorFlow always creates a temp file first (which unfortunately matches the given wildcard pattern, but this is not important here I think).
This temporary file is deleted soon, but if this happens when `tf.gfile.Glob` is being called, it crashes.
It crashes within the `tf.gfile.Glob` (i.e. not within the for-cycle in my code) with tensorflow.python.framework.errors_impl.NotFoundError, which is difficult to handle (I could catch the exception and try the same call again, but that's too hacky).
According to my extensive tests, `glob.glob` does not have this non-deterministic bug.

Reply:
the benefit of tf.gfile is that it will support many underlying filesystems, including Google cloud storage and internal filesystems. glob won't know what to do with a path that starts with gs://.

would updating the pattern to 'ckpt-[0-9]*.index' work?
i'm thinking maybe not because it's probably failing on a IsDirectory call.

i'm inclined to advocate for a retrying_glob function to maintain portability across filesystems though i agree it's hacky.

Reply:
I know the benefit of tf.gfile and this was the reason why I tried it first.
I would prefer a different solution:
* import tf.gfile wrapper once which overrides glob.glob, and keep glob.glob everywhere in the code
* make the wrapper detect special systems (`gs://`), but otherwise fallback to the standard (and reliable) `glob.glob`

If this is not acceptable, I will catch the exception and call again.
Even if the bug is rare, it is annoying when running many week-long experiments and some of the evaluation jobs crash.

> would updating the pattern to 'ckpt-[0-9]*.index' work

No. I use `stepfiles_iterator` also in `t2t-bleu`, where the files being iterated are not checkpoints, but translations of the test set, which I decided to have no extension (but `path_suffix` can be e.g. `.txt` as well). For this reason I call it `stepfiles` and not `checkpoints`.

Reply:
Let's have the retrying tf.gfile.Glob with the exception catching. Monkey-patching a lib would be really confusing for people reading the code and not understanding why something is happening.